core/src/main/java/org/apache/iceberg/stats/StatsReader.java

```diff
@@ -0,0 +1,177 @@
+
+      int footerPayloadSize = StatsFormat.readIntegerLittleEndian(footer, footerSize - 20);
+      Preconditions.checkState(footerPayloadSize == footerSize - 24,
+          "Unexpected footer payload size value %s for footer size %s", footerPayloadSize, footerSize);
+
+      FileMetadata readFileMetadata;
+      ByteBuffer footerPayload = ByteBuffer.wrap(footer, 4, footerPayloadSize);
+      if (!compressed) {
+        readFileMetadata = parseFileMetadata(footerPayload);
+      } else {
+        ByteBuffer footerJson = StatsFormat.decompressFooterPayload(footerPayload);
+        readFileMetadata = parseFileMetadata(footerJson);
+      }
+      this.knownFileMetadata = Optional.of(readFileMetadata);
+    }
+    return knownFileMetadata.get();
+  }
+
+  /**
+   * @throws IOException          when I/O error occurs
+   * @throws UncheckedIOException when I/O error occurs
+   */
+  // Note: The method is marked as throwing IOException to allow future implementation evolution.
+  // Currently, the exception isn't being thrown.
+  public Stream<Map.Entry<BlobMetadata, ByteBuffer>> readAll(List<BlobMetadata> blobs) throws IOException {
+    if (blobs.isEmpty()) {
+      return Stream.empty();
+    }
+
+    // TODO inspect blob offsets and coalesce read regions close to each other
+
+    return blobs.stream()
+        .sorted(Comparator.comparingLong(BlobMetadata::offset))
+        .map((BlobMetadata blobMetadata) -> {
+          try {
+            input.seek(blobMetadata.offset());
+            byte[] bytes = new byte[Math.toIntExact(blobMetadata.length())];
+            ByteStreams.readFully(input, bytes);
+            ByteBuffer data = ByteBuffer.wrap(bytes);
+            if (blobMetadata.compressionCodec() != null) {
+              StatsCompressionCodec codec = StatsCompressionCodec.forName(blobMetadata.compressionCodec());
+              data = StatsFormat.decompressBlob(codec, data);
+            }
+            return new SimpleEntry<>(blobMetadata, data);
+          } catch (IOException e) {
+            throw new UncheckedIOException(e);
+          }
+        });
+  }
+
+  private static void checkMagic(byte[] data, int offset) {
+    int read = StatsFormat.readIntegerLittleEndian(data, offset);
+    if (read != StatsFormat.MAGIC_AS_NUMBER_LE) {
+      throw new IllegalStateException(String.format(
+          "Invalid file: expected magic at offset %s (%s i.e. %s as integer) but got %s",
+          offset, Arrays.toString(StatsFormat.getMagic()), StatsFormat.MAGIC_AS_NUMBER_LE, read));
+    }
+  }
+
+  private static boolean isFlagSet(int setFlags, int testedFlag) {
+    return (setFlags & testedFlag) == testedFlag;
+  }
+
+  private int getFooterSize() throws IOException {
+    if (!knownFooterSize.isPresent()) {
+      int footerTailSize = Math.toIntExact(Math.min(fileLength, 20));
+      byte[] footerTail = readInput(fileLength - footerTailSize, footerTailSize);
+
+      checkMagic(footerTail, footerTailSize - 4);
```

Comment:
Like before I think it would be helpful if a lot of these magic numbers had Constants to label them.

Reply:
fixed